src/mbgl/text/collision_tile.cpp

```diff
@@ -149,48 +167,71 @@ Box CollisionTile::getTreeBox(const Point<float>& anchor, const CollisionBox& bo
 }
 
 std::vector<IndexedSubfeature> CollisionTile::queryRenderedSymbols(const GeometryCoordinates& queryGeometry, float scale) {
-
     std::vector<IndexedSubfeature> result;
-    if (queryGeometry.empty()) return result;
+    if (queryGeometry.empty() || (tree.empty() && ignoredTree.empty())) {
+        return result;
+    }
 
+    // Account for the rounding done when updating symbol shader variables.
+    float zoom = util::log2(scale);
+    zoom = std::ceil(zoom * 10.0f) / 10.0f;
+    scale = std::pow(2.0f, zoom);
+
+    // Generate a rotated geometry out of the original query geometry.
+    // Scale has already been handled by the prior conversions.
+    BGMPolygon polygon;
+    for (const auto& point : queryGeometry) {
+        auto rotated = util::matrixMultiply(rotationMatrix, convertPoint<float>(point));
+        boost::geometry::append(polygon, CollisionPoint { rotated.x, rotated.y });
+    }
+
+    // Predicate for ruling out already seen features.
     std::unordered_map<std::string, std::unordered_set<std::size_t>> sourceLayerFeatures;
+    auto seenFeature = [&] (const CollisionTreeBox& treeBox) -> bool {
+        const IndexedSubfeature& feature = std::get<2>(treeBox);
+        const auto& seenFeatures = sourceLayerFeatures[feature.sourceLayerName];
+        return seenFeatures.find(feature.index) == seenFeatures.end();
+    };
 
-    mapbox::geometry::multi_point<float> rotatedPoints {};
-    rotatedPoints.reserve(queryGeometry.size());
-    std::transform(queryGeometry.cbegin(), queryGeometry.cend(), std::back_inserter(rotatedPoints),
-                   [&](const auto& c) { return util::matrixMultiply(rotationMatrix, convertPoint<float>(c)); });
-    const auto box = mapbox::geometry::envelope(rotatedPoints);
+    // Check if feature is rendered (collision free) at current scale.
+    auto visibleAtScale = [&] (const CollisionTreeBox& treeBox) -> bool {
+        const CollisionBox& box = std::get<1>(treeBox);
+        return scale >= box.placementScale && scale <= box.maxScale;
+    };
 
-    const auto& anchor = box.min;
-    CollisionBox queryBox(anchor, 0, 0, box.max.x - box.min.x, box.max.y - box.min.y, scale);
-    auto predicates = bgi::intersects(getTreeBox(anchor, queryBox));
+    // Check if query polygon intersects with the feature box at current scale.
+    auto intersectsAtScale = [&] (const CollisionTreeBox& treeBox) -> bool {
+        const Box& box = std::get<0>(treeBox);
+        const auto anchor = bg::return_centroid<CollisionPoint>(box);
```

Comment:
Good point - at some point I attempted using `CollisionBox` but I was obtaining wrong values because I forgot to account for the `yStretch` multiplication. I'll modify the code to use `CollisionBox`.